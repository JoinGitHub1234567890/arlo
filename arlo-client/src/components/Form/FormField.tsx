import React from 'react'
import styled from 'styled-components'

const Wrapper = styled.div`
  width: 45%;
`

const Field = styled.input`
  width: 100%;
`

const ErrorLabel = styled.p`
  width: 100%;
  color: #ff0000;
  font-size: 10px;
`

interface Props {
  field?: any
  form?: any
  disabled?: boolean
  value?: string | number
  onChange?: (e: React.ChangeEvent<any>) => void
  onBlur?: (e: any) => void
  name?: string
  type?: string
  error?: string
  touched?: boolean
  className?: string
}

const FormField = ({
  field,
  form: { touched, errors },
  disabled,
  className,
  ...rest
}: Props) => (
  <Wrapper className={className}>
    <Field disabled={disabled} {...field} {...rest} />
    {errors[field.name] && touched[field.name] && (
      <ErrorLabel>{errors[field.name]}</ErrorLabel>
    )}
  </Wrapper>
)

export default FormField
