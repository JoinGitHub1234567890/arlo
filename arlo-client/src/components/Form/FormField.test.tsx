import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { FormikProps } from 'formik'
import FormField from './FormField'

describe('FormField', () => {
  it('renders a Field', () => {
    const field = {
      name: 'field-name',
      value: 'text value',
      onChange: jest.fn(),
      onBlur: jest.fn(),
    }
    const form = ({
      errors: {},
      touched: {},
    } as any) as FormikProps<any> // eslint-disable-line @typescript-eslint/no-object-literal-type-assertion
    const { container } = render(
      <FormField field={field} form={form} type="text" />
    )

    expect(container).toMatchSnapshot()
  })

  it('renders a numeric Field', () => {
    const field = {
      name: 'field-name',
      value: 'text value',
      onChange: jest.fn(),
      onBlur: jest.fn(),
    }
    const form = ({
      errors: {},
      touched: {},
      setFieldTouched: jest.fn(),
      setFieldValue: jest.fn(),
    } as any) as FormikProps<any> // eslint-disable-line @typescript-eslint/no-object-literal-type-assertion
    const { container, getByTestId } = render(
      <FormField field={field} form={form} type="number" data-testid="testid" />
    )

    expect(container).toMatchSnapshot()

    const input = getByTestId('testid')
    fireEvent.change(input, { target: { value: 'new value' } })
    fireEvent.blur(input)

    expect(field.onChange).toHaveBeenCalledTimes(0)
    expect(form.setFieldTouched).toHaveBeenCalledTimes(1)
    expect(form.setFieldValue).toHaveBeenCalledTimes(1)
  })

  it('renders a Field with errors', () => {
    const field = {
      name: 'field-name',
      value: '',
      onChange: jest.fn(),
      onBlur: jest.fn(),
    }
    const form = ({
      errors: {
        'field-name': 'Required',
      },
      touched: {
        'field-name': true,
      },
    } as any) as FormikProps<any> // eslint-disable-line @typescript-eslint/no-object-literal-type-assertion
    const { container } = render(
      <FormField field={field} form={form} type="text" />
    )

    expect(container).toMatchSnapshot()
  })

  it('calls onChange', () => {
    const field = {
      name: 'field-name',
      value: '',
      onChange: jest.fn(),
      onBlur: jest.fn(),
    }
    const form = ({
      errors: {},
      touched: {},
    } as any) as FormikProps<any>
    const { getByTestId } = render(
      <FormField field={field} form={form} type="text" data-testid="testid" />
    )

    fireEvent.change(getByTestId('testid'), { target: { value: 'new value' } })

    expect(field.onChange).toBeCalledTimes(1)
  })
})
