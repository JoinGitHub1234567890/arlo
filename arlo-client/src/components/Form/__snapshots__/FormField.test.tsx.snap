// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField renders a Field 1`] = `
<div>
  <div
    class="sc-bdVaJa NQGah"
  >
    <input
      class="sc-bwzfXH jwfMIB"
      name="field-name"
      type="text"
      value="text value"
    />
  </div>
</div>
`;

exports[`FormField renders a Field with errors 1`] = `
<div>
  <div
    class="sc-bdVaJa NQGah"
  >
    <input
      class="sc-bwzfXH jwfMIB"
      name="field-name"
      type="text"
      value=""
    />
    <p
      class="sc-htpNat bhgSmC"
      data-testid="field-name-error"
    >
      Required
    </p>
  </div>
</div>
`;
